### ISSUE TITLE
complex hard constraint

### ISSUE THREAD

**Issue by HHHMX** [lululxvi/deepxde#90]
@lululxvi 

hello, Lu

      i try to use deepxde to solve the navier stokes as follow
![1B798E05-98E4-4ADB-8312-8DA7C2E2930B](https://user-images.githubusercontent.com/66006577/87869353-ac0b7000-c9d1-11ea-9cdf-f5c868596bac.jpeg)

   the bc is that in the right, left and lower wall,   veloctiy u and v is zero.

 and in the upper wall, the velocity u is 1m/s, v is zero,

as for the pessure, the boundary condition is that in left and right wall, dp/dx = 0, in lower and upper wall, 

dp/dy = 0 

but the result of trainning didnt satisfy  the boundary , so i want to ues the hard constraints,

for my issuse, the hard constraints should be 

                        v(x, y)  = x * y * (x-d) * (y-d) * v,    u(x,y) = x * y * (x-d)  * u(x,y)

                        p =   ( ( x * (x-d) )**2 + ( y * (y-d) )**2   ) * p

(u(x, d) = 1 cant be satisfyed in the sametime, so i would use soft constraints to satisfy this bc)

but i dont know how to apply these by ues net.outputs_modify,  and i didnt find some uesful information .

**Comment by lululxvi** [comment]
- If the result of training didn't satisfy the boundary, you can use a larger weight for the BC loss.
- Use hard constraints by modifying the network output, e.g., v:
```python
def modify_output(X, y):
    x, y = X[;, 0:1], X[:, 1:2]
    u, v, p = y[:, 0:1], y[:, 1:2], y[2:3]  # Assume the order of output is u, v, p
    u_new = x * y * (x - d)  * u
    v_new = x * y * (x - d) * (y - d) * v
    p_new = ((x * (x - d)) ** 2 + (y * (y - d)) ** 2) * p
    return tf.concat((u_new, v_new, p_new), axis=1)

net.outputs_modify(modify_output)
```

**Comment by HHHMX** [comment]
thanks for your advice and the great deepxde,!    I have got  a good result .

**Comment by SanjuSoni** [comment]
> * If the result of training didn't satisfy the boundary, you can use a larger weight for the BC loss.
> * Use hard constraints by modifying the network output, e.g., v:
> 
> ```python
> def modify_output(X, y):
>     x, y = X[;, 0:1], X[:, 1:2]
>     u, v, p = y[:, 0:1], y[:, 1:2], y[2:3]  # Assume the order of output is u, v, p
>     u_new = x * y * (x - d)  * u
>     v_new = x * y * (x - d) * (y - d) * v
>     p_new = ((x * (x - d)) ** 2 + (y * (y - d)) ** 2) * p
>     return tf.concat((u_new, v_new, p_new), axis=1)
> 
> net.outputs_modify(modify_output)
> ```

Dear @lululxvi 
the value of u_new at y = d is different from the original boundary condition.

i.e.` u_new(x, d) = x*d*(x-d)*1 `
which is not same as u_x = 1.

if I am wrong kindly suggest  me.

**Comment by lululxvi** [comment]
I am not sure... I just demonstrate how to use hard constraints.

**Comment by SanjuSoni** [comment]
Thanks, @lululxvi and kindly suggest me for my BC scenario mentioned below.

I am also trying to enforce the BC over the square domain.

a square domain has BC as:
psi = 0    on all sides of the unit square.  i.e. x = 0, 1     y = 0,  1

T  = 1 on x = 0
T = 0 on x = 1


I have coded it as: 

```
net.apply_output_transform(
        lambda x, y : tf.concat(
            [ x[:, 0:1] * x[:,1:2]*(1-x[:,0:1])*(1-x[:,1:2])*y[:,0:1],
            (1-x[:,0:1])*x[:,1:2]*y[:,1:2]], axis =1
            )   )
```

FYI:
x[:, 0:1]   is x
x[:, 1:2]   is y
y[:, 0:1]   is S
y[:, 1:2]  is T


Kindly suggest me, for the above code is right for BC strict enforcement.

**Comment by lululxvi** [comment]
The usage is right, but the math may be not. Does `(1-x[:,0:1])*x[:,1:2]*y[:,1:2]` satisfy `T = 1 on x = 0`?

**Comment by SanjuSoni** [comment]
> The usage is right, but the math may be not. Does `(1-x[:,0:1])*x[:,1:2]*y[:,1:2]` satisfy `T = 1 on x = 0`?

at the right side of domain, i.e. at x =1,   T_new = 0 satisfy that expression.
but at x = 0,   T_new = 1*y*1  which is varying from 0 to 1.

Dear @lululxvi 
       can we use T_new = cos((pi/2)*x)) * T

because it satisfies the BC for T at x=0 and x =1 

**at x = 0,                    T_new = 1
and at x= 1,              T_new = 0**
   
for that should I go for `net.output_modify(...)`  or `net.apply_output_transform(...)`

```
net.apply_output_transform(
        lambda x, y : tf.concat(
            [     
            x[ :, 0:1 ] * x[ :, 1:2 ]*(1 - x[ :, 0:1 ])*( 1 - x[ :, 1:2 ] )*y[ :, 0:1],
            np.cos( (3,14 / 2) * x[ : , 1] ) * y[ :, 1:2] 
            ], axis =1
            )   )
```


please comment on it.

**Comment by lululxvi** [comment]
- Does `cos((pi/2)*x)) * T` satisfy `T = 1 on x = 0`?
- I may use something like `x(1-x)T + 1 - x`. Of course, the solution is not unique.
- Use the new API `apply_output_transform`

**Comment by SanjuSoni** [comment]
Thanks @lululxvi 
That is also a correct expression.

I have done this coding to my problem as:

```

    net.apply_output_transform(
        lambda x, y : tf.concat(
            [ x[:, 0:1] * x[:,1:2]*(1-x[:,0:1])*(1-x[:,1:2])*y[:,0:1],
            x[:,0:1]*(1-x[:, 0:1])*y[:,1:2]+ 1 - x[:,0:1]], axis = 1
            )   )
```

but, the training process does not look good, because few boundary conditions are not improving after iteration.

![training_triangle](https://user-images.githubusercontent.com/61820415/91192481-9c3f2400-e713-11ea-846b-1199a992747d.JPG)

also, can you suggest to me, is my GPU is compatible with the DEEPXDE module?


I have attached my code below:

 
[BC_Enforced_Darcy.zip](https://github.com/lululxvi/deepxde/files/5124619/BC_Enforced_Darcy.zip)

**Comment by lululxvi** [comment]
1000 iterations cannot say anything. Maybe the loss terms are not balance. see "Q: I failed to train the network or get the right solution, e.g., the training loss is large." at FAQ for lots of possible solutions.

**Comment by SanjuSoni** [comment]
Thanks for your suggestions @lululxvi ,
But I run it for 15000 epochs due to time crunch I had taken this snapshot for 2000 epochs.

Also, I plotted the result which is better than previous one and I would improve it further based on your suggestions.

Thanks @lululxvi

**Comment by SanjuSoni** [comment]
@lululxvi  Sir,
The above discussion has been done for the case of Dirichlet BC.

in case of Neumann BC, imposed on the boundary, than how can we impose the hard constriant.

In my case:

dT/dx = 0 , for y = 0, and y =1

T = 1,   for x =0
T = 0, for x = 1

S = 0, x = 0, 1, and y = 0, 1

kindly suggest me for the case of Neumann BC

**Comment by lululxvi** [comment]
Neumann is tricky. See the paper [Artificial Neural Networks for Solving Ordinary and Partial Differential Equations, 1998] on how to define `apply_output_transform`. But I suggest not to enforce Neumann, but use standard loss (maybe with a larger weight).

**Comment by smao-astro** [comment]
> @lululxvi Sir,
> The above discussion has been done for the case of Dirichlet BC.
> 
> in case of Neumann BC, imposed on the boundary, than how can we impose the hard constriant.
> 
> In my case:
> 
> dT/dx = 0 , for y = 0, and y =1
> 
> T = 1, for x =0
> T = 0, for x = 1
> 
> S = 0, x = 0, 1, and y = 0, 1
> 
> kindly suggest me for the case of Neumann BC

Also see [Artificial Neural Network Method for Solution of Boundary Value Problems With Exact Satisfaction of Arbitrary Boundary Conditions
](https://ieeexplore.ieee.org/document/5061501)

I agree with @lululxvi that hard Neumann is tricky, because 

- you might suffer from stability issue of the solution
- Since now you have more gradient to compute, it need more computational resources

**Comment by SanjuSoni** [comment]
Thanks, @lululxvi  and @smao-astro 
So, can we say that only the Dirichlet BC implementation may serve the applicability of Neumann BC also?
And, the deep-learning result will not deviate from its original results.

**Comment by lululxvi** [comment]
Ideally, all BCs can be treated as a loss term, if you choose the correct hyperparameters, e.g., weights, etc. But in practice, using hard constrain may make the learning of NN easier. Dirichlet BC is easy to implement, and Neumann BC is relatively hard to implement. You can also use hard constrain for Neumann BC if you can get the formula easily. But this is my experience, and I don't have a conclusion for you. (By the way, I even have a case where soft constrain is better than hard constrain for Dirichlet BC)

**Comment by SanjuSoni** [comment]
Thanks for your valuable comments. I'll follow your suggestion.

**Comment by voloddia** [comment]
I'm trying to enforce these conditions: 
`y(0, t) = 1
y(x, 0) = 0`
With `net.apply_output_transform(lambda x, y: x[:, 0:1]*x[:, 1:2]*y + tf.sign(x[:, 1:2]))`, these conditions are satisfied but then domain points are all 0 (I guess due to derivative of `tf.sign`). 
Please suggest me about this.

**Comment by lululxvi** [comment]
You requirement is ill-posed. For example, what is `y(0, 0)`?

**Comment by voloddia** [comment]
Yes, I realized this issue. 
Currently I assume: 
`y(0, t>0) = 1`
`y(x>=0, 0) = 0`
The solution suggests: 
`y(0, t>=0) = 1`
`y(x>0, 0) = 0`
But in either way, I haven't figure out how to enforce them.

**Comment by lululxvi** [comment]
In both cases, you solution is discontinuous, which could be hard to solve. `tf.sign` seems correctly, but maybe there is something wrong in computing the derivative in Tensorflow (I am not sure about this). You may only enforce one of them by `apply_output_transform`.

**Comment by Laoliu66** [comment]
> 感谢您的建议和伟大的 deepxde，！我得到了一个很好的结果。


Hello, I have encountered difficulties in a similar problem, can you share your code?

**Comment by MINE0126** [comment]
Hello, @lululxvi @HHHMX 
How to write this hard constraint expression? Please give me some ways to describe expressions for hard constraints! Thank you very much!
For example
![2-D heat equation](https://user-images.githubusercontent.com/107633173/189511860-b4c57bbe-29bc-4940-8bb8-9a1acecf4501.png)
The Boundary conditions 1: {T=0.8 on (y=0.8 and x=[0,0.8])} and the Boundary conditions 2: {T=0.2 on (y=0.8 and x=[1.6,2.4])}， I don't know how to write their expressions and how to describe them using "apply_output_transform". Can you give me some advice?
Looking forward to your reply？

**Comment by lululxvi** [comment]
Check Section 2.3 at https://doi.org/10.1137/21M1397908. You need to come up with the l(x) and g(x). But it doesn't seem straightforward for your case.
